bundle/edu.gemini.pot/src/main/java/edu/gemini/spModel/template/InstantiationFunctor.java

```diff
@@ -34,29 +42,29 @@ public class InstantiationFunctor extends DBAbstractFunctor {
 
     private static final Logger LOGGER = Logger.getLogger(InstantiationFunctor.class.getName());
 
-    private final Map<ISPTemplateGroup, Set<ISPTemplateParameters>> selection = new HashMap<>();
+    private ListMap<ISPTemplateGroup, ListSet<ISPTemplateParameters>> selection = ListMap$.MODULE$.empty();
 
     /**
      * Add a group/params pair to the set of instantiations to be performed.
      * @param group a template group
      * @param params a set of params
      */
     public void add(ISPTemplateGroup group, ISPTemplateParameters params) {
-        Set<ISPTemplateParameters> list = selection.get(group);
-        if (list == null) {
-            list = new HashSet<>();
-            selection.put(group, list);
-        }
-        list.add(params);
+        final Option<ListSet<ISPTemplateParameters>> optSetOfParams = selection.get(group);
+
+        final ListSet<ISPTemplateParameters> setOfParams =
+          optSetOfParams.isEmpty() ? ListSet$.MODULE$.empty() : optSetOfParams.get();
+
+        selection = selection.updated(group, setOfParams.$plus(params));
     }
 
     public void execute(IDBDatabaseService db, ISPNode node, Set<Principal> principals) {
         try {
             List<ISPGroup> newGroups = new ArrayList<>();
             final ISPProgram prog = db.lookupProgram(node.getProgramKey());
-            for (Map.Entry<ISPTemplateGroup, Set<ISPTemplateParameters>> e : selection.entrySet()) {
-                for (ISPTemplateParameters ps : e.getValue()) {
-                    final ISPTemplateGroup templateGroup = e.getKey();
+            for (Tuple2<ISPTemplateGroup, ListSet<ISPTemplateParameters>> e : WrapAsJava$.MODULE$.asJavaIterable(selection.toList())) {
+                for (ISPTemplateParameters ps : WrapAsJava$.MODULE$.asJavaIterable(e._2.reversed())) {
```

Comment:
Oddly, I have to reverse the `Set` of parameters.  New parameters are added using the `+` method (translated to `.$plus` in Java).  I confirmed they `add` method is called in the expected order so `+` must be prepending instead of appending?